tools/browser_compat_data_update.py

```diff
@@ -0,0 +1,113 @@
+                unknown_keys = set(support.keys()) - {
+                    "version_added",
+                    "notes",
+                    "alternative_name",
+                }
+                if unknown_keys:
+                    # Assume unsupported if feature has been removed after it was added, requires flags or something like that.
+                    continue
+                out_value[browser] = parse_version(version_added.lstrip("≤"), browser)
+            # Don't bother adding JS features that have ~always been supported, we're not going to feature-test those.
+            if not all(
+                min_interesting_version >= out_value.get(browser, maxsize)
+                for browser, min_interesting_version in min_interesting_versions.items()
+            ):
+                out["#"] = out_value
+    return out
+
+
+out = {}
+
+# Pull JavaScript builtins support from MDN's Browser Compat Data.
+data = requests.get("https://unpkg.com/@mdn/browser-compat-data/data.json").json()
+# Fill minimum browser versions based on year.
+min_interesting_versions = {
+    browser: min(
+        parse_version(version, browser)
+        for version, release in data["browsers"][browser]["releases"].items()
+        if "release_date" in release and int(release["release_date"][:4]) >= min_interesting_year
+    )
+    for browser in interesting_browsers
+}
+print("Minimum interesting versions: ", min_interesting_versions)
+out["js"] = traverse_bcd(data["javascript"]["builtins"])
+# There are a lot of browser APIs and very few we care about, so only pull them to keep JSON small.
+for api in ["AudioWorklet"]:
+    # Just add them to the same JS namespace.
+    out["js"][api] = traverse_bcd(data["api"][api])
+# Manually add 'performance' because browser-compat-data currently doesn't match https://caniuse.com/high-resolution-time.
+# It shows when API is available, but not when it returns actually high-resolution time.
+out["js"]["performance"] = {
+    "#": {"chrome": 24, "firefox": 15, "nodejs": 160000, "safari": 80000}
+}
+
+# BCD data is not as complete for Wasm features as webassembly.org, at least for now.
+data = requests.get("https://webassembly.org/features.json").json()
+out_wasm = out["wasm"] = {}
+for feature in data["features"]:
+    out_wasm[feature] = {"#": {}}
+for browser in ["Chrome", "Firefox", "Safari", "Node.js"]:
+    normalized_browser = browser.lower().replace(".", "")
+    for feature, support in data["browsers"][browser]["features"].items():
+        if type(support) is list:
+            # If support is a list, first item is the version and 2nd is a text note.
+            support = support[0]
+        if type(support) is not str or support == "flag":
+            continue
+        out_wasm[feature]["#"][normalized_browser] = parse_version(
+            support, normalized_browser
+        )
+
+with open(Path(__file__).parent / "browser_compat_data.json", "w") as f:
```

Comment:
Maybe use `utils.write_file` here instead of the with block?

Reply:
That would mean stringifying JSON in-memory first instead of writing to file in streaming way? Probably not a big difference, just wanted to point out it's not exactly the same.

Reply:
Limiting the size of the json should make that unimportant.

Reply:
One problem I'm having when trying to import utils is:

```shell
> py .\tools\browser_compat_data_update.py
Traceback (most recent call last):
  File "C:\Users\me\Documents\emscripten\tools\browser_compat_data_update.py", line 5, in <module>
    from utils import write_file
  File "C:\Users\me\Documents\emscripten\tools\utils.py", line 12, in <module>
    from . import diagnostics
ImportError: attempted relative import with no known parent package
```

What am I doing wrong?

Reply:
See `tools/wasm-sourcemap.py` for an example:

```
__scriptdir__ = os.path.dirname(os.path.abspath(__file__))                                              
__rootdir__ = os.path.dirname(__scriptdir__)                                                                                                                                
sys.path.insert(0, __rootdir__)  
```

Then `from tools import utils`

Reply:
Also, perhaps this new tool should live in `tools/maint`?

Reply:
> See `tools/wasm-sourcemap.py` for an example:

Hm is all of that really worth it just to avoid the `with` block?

> Also, perhaps this new tool should live in tools/maint?

I'm not very familiar with the folder structure here, but sure, I can move it.

Reply:
> > See `tools/wasm-sourcemap.py` for an example:
> 
> Hm is all of that really worth it just to avoid the `with` block?

probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.

> 
> > Also, perhaps this new tool should live in tools/maint?
> 
> I'm not very familiar with the folder structure here, but sure, I can move it.

I just noticed there are some other scripts that can/should be moved there too.

Reply:
> probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.

But some other scripts like create_entry_points already don't do that, and also opt for `with` blocks. I thought that was intentional choice to avoid similar "glue".

Reply:
> > probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.
> 
> But some other scripts like create_entry_points already don't do that, and also opt for `with` blocks. I thought that was intentional choice to avoid similar "glue".

I don't feel strongly but over time it would be good to make all this consistent.

For what is worth I just add the sys.path boilerplate to create_entry_points as part of #20586

Reply:
I'm not that deeply familiar with Python, but is there a way to declare all those scripts as a single package or something so it automatically allows such relative imports without duplicating the code? (figured I'd ask since the error talks about "parent package")

Reply:
Its complicated because when you run a script like that its running as `__main__` and not considered part of the `tools` package at that point.   It sadly doesn't just work as you might hope.  

You need to add the emscripten root to the sys.path as if the script was run from the top level (like emcc.py is).

Unless you want to go down the rabbit hole I suggest you just copy those three lines which update `sys.path` accordingly.

Reply:
Basically every script that wants to be directly runnable needs to do that little dance.

Ideally the PYTHON_PATH would already be setup to point the top level directly by the launching process..  but in this case we kind of want these files themselves to be directly runnable without creating extra launcher scripts.

Reply:
That's a shame. Thanks for the explanation.